ribbon.py

```diff
@@ -0,0 +1,545 @@
+        if marker is not None and ribbon is not None:
+            image_offset = (320-self.MARKER_CROP_WIDTH)/2
+            marker_x = marker[0] + image_offset
+            ribbon_x = ribbon[0]
+            self.last_marker_time = time.time()
+            if (marker_x > ribbon_x) == self.MARKERS_ON_THE_LEFT:
+                 #if the markers are the same side as they're meant to be, we're going the right way
+                 direction = True
+            else:
+                 direction = False
+            if direction == self.MARKERS_ON_THE_LEFT:
+                logger.info ("marker spotted on the left")
+            else:
+                logger.info ("marker spotted on the right")
+        else:
+            #if either marker or ribbon can't be seen, assume we're ok
+            logger.info ("marker spotted but no ribbon")
+            direction = True
+        return direction
+
+    def stuck(self):
+        #if its been more than the timeout since we last saw a marker, we're probably stuck
+        if (self.last_marker_time + self.MARKER_TIMEOUT) < time.time() or ((self.last_t_error == self.last_before_that_t_error) and self.last_t_error is not None and self.last_t_error <> 0 and not self.isstuck):
+            self.isstuck = True
+        else:
+            self.isstuck = False
+        return False #self.isstuck
+
+    def escape(self):
+        print "escaping"
+        if random.choice([True, False]):
+            self.drive.move(0, self.ESCAPE_SPEED)
+        else:
+            self.drive.move(0, -self.ESCAPE_SPEED)
+        time.sleep(self.ESCAPE_TIME)
+        self.drive.move(0, 0)
+        #reset timeout
+        self.last_marker_time = time.time()
+
+    def turn_around(self):
+        print "marker wrong side of ribbon, turning around"
+        self.drive.move(self.TURN_AROUND_SPEED, 0)
+        time.sleep(self.TURN_AROUND_TIME)
+        self.drive.move(0, 0)
+
+    def ribbon_following(self, marker_image, ribbon_image, image):
+        ribbon_x, ribbon_y, ribbon_area, ribbon_contour = find_largest_contour(ribbon_image)
+        if ribbon_area > self.MIN_CONTOUR_AREA:
+            ribbon = [ribbon_x, ribbon_y, ribbon_area, ribbon_contour]
+            image = cv2.cvtColor(image, cv2.COLOR_RGB2HSV)
+#            print colour_of_contour(image, ribbon_contour)
+        else:
+            ribbon = None
+        pygame.mouse.set_pos(ribbon_y, 320 - ribbon_x)
+        marker_x, marker_y, marker_area, marker_contour = find_largest_contour(marker_image)
+        if marker_area > self.MIN_MARKER_AREA:
+            marker = [marker_x, marker_y, marker_area, marker_contour]
+        else:
+            marker = None
+        if marker:
+            self.mode_number = 1
+            self.drive.move(0,0)
+            time.sleep(0.6)
+            self.marker(marker_image, ribbon_image, image)
+        if self.tracking:
+            if not self.stuck():
```

Comment:
if the `stuck()` method is legit hardcoded to be returning `False` then you can ditch this conditional (and the `self.stuck()` method, as long as it isn't used elsewhere)